### PR TITLE
fuzz: improve `scriptpubkeyman` target

### LINKED ISSUE #30541: scriptpubkeyman fuzz target TopUp is slow (2/N)
Follow-up to https://github.com/bitcoin/bitcoin/issues/30476

Found by OSS-Fuzz / https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=70780 

Seems wasteful to spend time in `TopUp` while fuzzing, so my recommendation would be to limit TopUp iterations a bit.

Context: [clusterfuzz-testcase-scriptpubkeyman-5934021442338816.bin.txt](https://github.com/user-attachments/files/16409466/clusterfuzz-testcase-scriptpubkeyman-5934021442338816.bin.txt)

```
$ FUZZ=scriptpubkeyman perf record -g --call-graph dwarf  ./src/test/fuzz/fuzz -runs=1 ./clusterfuzz-testcase-scriptpubkeyman-5934021442338816.bin.txt
$ hotspot  ./perf.data 
```

Screenshot from 2024-07-29 10-05-27

### PR DESCRIPTION
**Pull request by brunoerg**
Fixes #30541

This PR aims to improve `scriptpubkeyman` target to avoid timeouts. The input provided in #30541 takes too much time to run because it basically calls only `MarkUnusedAddresses` (300 times * number of spks). The following changes were made to improve it:

- Reduce keypool size.
- When calling `MarkUnusedAddresses`, do it with one of the spks per iteration.
- Remove the specific `AddDescriptorKey` call since it is already covered with `AddWalletDescriptor`.
- Limit number of iterations to a reasonable value.

### CODE REVIEW

**Comment by brunoerg** [comment]
cc: @maflcko

**Review by maflcko — COMMENTED** [review]
> calls only `MarkUnusedAddresses` (300 times * number of spks)

Are they different spks? IIRC when I looked, they looked all identical, which I am not sure is working as intended.

File: src/wallet/test/fuzz/scriptpubkeyman.cpp [hunk 1]
```diff
@@ -103,7 +103,20 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
...
+        if (!wallet_desc.has_value()) {
+            good_data = false;
+            return;
+        }
+        std::string error;
+        if (spk_manager->CanUpdateToWalletDescriptor(wallet_desc->first, error)) {
+            auto new_spk_manager{CreateDescriptor(wallet_desc->first, wallet_desc->second, wallet)};
+            if (new_spk_manager != nullptr) spk_manager = new_spk_manager;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp:117]: You're right, there is no need for good_data here. It was a leftover. Gonna change it.

[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: Done.

**Comment by brunoerg** [comment]
> Are they different spks? IIRC when I looked, they looked all identical, which I am not sure is working as intended.

For that input, it was calling `MarkUnusedAddresses` with all spks every iteration. Besides that, every time `MarkUnusedAddresses` is called, the number of spks increases. That's why it takes a lot time to run.

Now, it will just call `MarkUnusedAddresses` with just one spk (randomly) per iteration. Besides that, the number of iterations was reduced for a reasonable value.

**Review by maflcko — APPROVED** [review]
lgtm ACK fd7cd3b8b24ad86dae5c16addbb5acafa18efc0a

File: src/wallet/test/fuzz/scriptpubkeyman.cpp [hunk 2]
```diff
@@ -147,26 +147,10 @@ FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
                 for (const CScript& spk : spks) {
                     if (fuzzed_data_provider.ConsumeBool()) {
                         spk_manager->MarkUnusedAddresses(spk);
+                        break;
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: Done

**Comment by brunoerg** [comment]
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/30563#discussion_r1700211699

**Comment by achow101** [comment]
ACK 401cc4ec70d67ba2aa0e078d2fab214e1c40742c

**Pull request merged**